beacon_node/operation_pool/src/attester_slashing.rs

```diff
@@ -0,0 +1,71 @@
+use crate::max_cover::MaxCover;
+use std::collections::HashMap;
+use types::{AttesterSlashing, BeaconState, ChainSpec, EthSpec};
+
+pub struct AttesterSlashingMaxCover<'a, T: EthSpec> {
+    slashing: &'a AttesterSlashing<T>,
+    effective_balances: HashMap<u64, u64>,
+}
+
+impl<'a, T: EthSpec> AttesterSlashingMaxCover<'a, T> {
+    pub fn new(
+        slashing: &'a AttesterSlashing<T>,
+        state: &BeaconState<T>,
+        spec: &ChainSpec,
+    ) -> Option<Self> {
+        let length = slashing.attestation_1.attesting_indices.len()
+            + slashing.attestation_2.attesting_indices.len();
+        let mut effective_balances: HashMap<u64, u64> = HashMap::with_capacity(length);
+
+        for vd in &slashing.attestation_1.attesting_indices {
+            let eff_balance = state.get_effective_balance(*vd as usize, spec).ok()?;
+            effective_balances.insert(*vd, eff_balance);
+        }
+
+        for vd in &slashing.attestation_2.attesting_indices {
+            let eff_balance = state.get_effective_balance(*vd as usize, spec).ok()?;
+            effective_balances.insert(*vd, eff_balance);
+        }
+
+        Some(Self {
+            slashing,
+            effective_balances,
+        })
+    }
+}
+
+impl<'a, T: EthSpec> MaxCover for AttesterSlashingMaxCover<'a, T> {
+    /// The result type, of which we would eventually like a collection of maximal quality.
+    type Object = AttesterSlashing<T>;
+    /// The type used to represent sets.
+    type Set = HashMap<u64, u64>;
+
+    /// Extract an object for inclusion in a solution.
+    fn object(&self) -> AttesterSlashing<T> {
+        self.slashing.clone()
+    }
+
+    /// Get the set of elements covered.
+    fn covering_set(&self) -> &HashMap<u64, u64> {
+        &self.effective_balances
+    }
+    /// Update the set of items covered, for the inclusion of some object in the solution.
+    fn update_covering_set(
+        &mut self,
+        best_slashing: &AttesterSlashing<T>,
+        covered_validator_indices: &HashMap<u64, u64>,
+    ) {
+        if best_slashing == self.slashing {
+            return;
+        }
```

Comment:
What's the reasoning behind this early return? I think the max coverage solver already takes care of not updating the `best_slashing` with its own set, and even if it does, it should just set its quality to 0 (which is fine).

Reply:
You're right max cover does and the reasoning for that, is not to waste time doing something that won't affect the outcome I thought? Meaning, if it's not going to be visiting this object again, why bother removing those validators from an object we're already using.  

The way I understood it and maybe i'm wrong, is that we want to touch as many validator indices with our two selected attester slashings?

Reply:
> Meaning, if it's not going to be visiting this object again, why bother removing those validators from an object we're already using.

Yeah, exactly. The max cover does this by marking the item as `available = false`, and then only selecting from items with `available = true`.